compiler/rustc_session/src/session.rs

```diff
@@ -232,6 +232,15 @@ pub trait SessionDiagnostic<'a, T: EmissionGuarantee = ErrorGuaranteed> {
     fn into_diagnostic(self, handler: &'a Handler) -> DiagnosticBuilder<'a, T>;
 }
 
+impl<'a, T: SessionDiagnostic<'a>> SessionDiagnostic<'a> for Spanned<T> {
```

Comment:
Maybe `T: AddToDiagnostic` (or whatever the `SessionSubdiagnostic` is called, we're in the process of changing it so I forget), then you can do:

```
/* approximately, it'll look something like this */
let diag = handler.create_diag(/* whatever the derive usually does */);
diag.set_primary_span(self.span);
self.node.add_to_diagnostic(diag);
diag
```

and then implement `SessionSubdiagnostic`'s trait for `LayoutError`?

Reply:
I don't think I want to implement `AddSubdiagnostic` for `LayoutError` because the main diagnostic should come from `LayoutError`. I've updated the commit with the lifetime error I referenced orgininally. To me it sounds like `self` doesn't live long enough.

Reply:
What if you implemented `Diagnostic` for `LayoutError` (without a `#[primary_span]`) and then implemented `Diagnostic` for `Spanned<T>` where `T: Diagnostic` and the impl for that was just:

```rust
/* approximately, it'll look something like this */
let Spanned { node, span } = self;
let diag = node.into_diagnostic(handler);
diag.set_primary_span(span);
diag
```

That way a `Spanned<T>` could be used to take any non-spanned diagnostic and make it spanned?